### PR TITLE
Migrate Button Block tests to Playwright

### PR DESCRIPTION
**Pull request by pooja-muchandikar**
## What?

Part of https://github.com/WordPress/gutenberg/issues/38851.

Migrate[ button-test.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-tests/specs/editor/blocks/buttons.test.js) to its Playwright version.

## Why?

Part of https://github.com/WordPress/gutenberg/issues/38851.

## How?

See[ MIGRATION.md](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/MIGRATION.md) for migration steps.

## Testing Instructions

Run` npm run test-e2e:playwright test/e2e/specs/editor/blocks/button.spec.js`

### CODE REVIEW

**Comment by pooja-muchandikar** [comment]
Hi @kevin940726 👋 

Please help me in reviewing this PR!!

Have a nice day ahead!

File: test/e2e/specs/editor/blocks/buttons.spec.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		editor,
+		page,
+		pageUtils,
+	} ) => {
+		// Regression: https://github.com/WordPress/gutenberg/pull/19885
+		await editor.insertBlock( { name: 'core/buttons' } );
+		await pageUtils.pressKeyWithModifier( 'primary', 'k' );
+		await page.waitForFunction(
+			() => !! document.activeElement.closest( '.block-editor-url-input' )
+		);
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: I know this is from the original code but it is a bit hacky and not accessible. Let's change this to role selectors and use Playwright-style assertions.

```suggestion
		await expect( page.locator( 'role=combobox[name="URL"i]' ) ).toBeFocused();
```

File: test/e2e/specs/editor/blocks/buttons.spec.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+		await page.keyboard.press( 'Escape' );
+		await page.waitForFunction(
+			() =>
+				document.activeElement ===
+				document.querySelector( '.block-editor-rich-text__editable' )
+		);
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: Same here:

```suggestion
		await expect( page.locator( 'role=textbox[name="Button text"i]' ) ).toBeFocused();
```

File: test/e2e/specs/editor/blocks/buttons.spec.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		editor,
+		page,
+		pageUtils,
+	} ) => {
+		// Regression: https://github.com/WordPress/gutenberg/issues/34307
+		await editor.insertBlock( { name: 'core/buttons' } );
+		await pageUtils.pressKeyWithModifier( 'primary', 'k' );
+		await page.waitForFunction(
+			() => !! document.activeElement.closest( '.block-editor-url-input' )
+		);
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: Ditto.

```suggestion
		await expect( page.locator( 'role=combobox[name="URL"i]' ) ).toBeFocused();
```

File: test/e2e/specs/editor/blocks/buttons.spec.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		await page.keyboard.type( 'https://example.com' );
+		await page.keyboard.press( 'Enter' );
+		await page.waitForFunction(
+			() =>
+				document.activeElement ===
+				document.querySelector(
+					'.block-editor-link-control a[href="https://example.com"]'
+				)
+		);
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: Ditto.

```suggestion
		await expect( page.locator( 'role=link[name=/^https:\\/\\/example\\.com]' ) ).toBeFocused();
```

File: test/e2e/specs/editor/blocks/buttons.spec.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+		await page.keyboard.press( 'Escape' );
+
+		// Focus should move from the link control to the button block's text.
+		await page.waitForFunction(
+			() =>
+				document.activeElement ===
+				document.querySelector( '[aria-label="Button text"]' )
+		);
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: Ditto.

```suggestion
		await expect( page.locator( 'role=textbox[name="Button text"i]' ) ).toBeFocused();
```

File: test/e2e/specs/editor/blocks/buttons.spec.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+		// The link control should still be visible when a URL is set.
+		const linkControl = await page.locator( '.block-editor-link-control' );
+		expect( linkControl ).toBeTruthy();
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: Ditto.

```suggestion
		await expect( page.locator( 'role=link[name=/^https:\\/\\/example\\.com/]' ) ).toBeVisible();
```

File: test/e2e/specs/editor/blocks/buttons.spec.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+		await editor.insertBlock( { name: 'core/buttons' } );
+		await page.keyboard.type( 'WordPress' );
+		await pageUtils.pressKeyWithModifier( 'primary', 'k' );
+		await page.keyboard.type( 'https://www.wordpress.org/' );
+		await page.keyboard.press( 'Enter' );
+		// Make sure that the dialog is still opened, and that focus is retained
+		// within (focusing on the link preview).
+		await page.waitForSelector(
+			':focus.block-editor-link-control__search-item-title'
+		);
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: Ditto.

```suggestion
		await expect( page.locator( 'role=link[name=/^https:\\/\\/wordpress\\.org/]' ) ).toBeFocused();
```

**Review by kevin940726 — COMMENTED** [review]
Thank you for making this PR! Some suggestions about using role selectors here and there but overall it looks good 👍 !

File: test/e2e/specs/editor/blocks/buttons.spec.js (hunk 4 above)
Comment (pooja-muchandikar) [reply test/e2e/specs/editor/blocks/buttons.spec.js]: Hi @kevin940726, Tried to change the selector as per suggestion but it is showing syntax error in regex. Tried to modify it but it is still showing parsing error. Could you please check on this?

image

Comment (kevin940726) [reply test/e2e/specs/editor/blocks/buttons.spec.js]: Ahh, yes, I was missing the `/` in the end.

```suggestion
		await expect( page.locator( 'role=link[name=/^https:\\/\\/example\\.com/]' ) ).toBeFocused();
```

Comment (pooja-muchandikar) [reply test/e2e/specs/editor/blocks/buttons.spec.js]: @kevin940726 

Updated, but still seems some issue, am I missing something?

image

Comment (kevin940726) [reply test/e2e/specs/editor/blocks/buttons.spec.js]: Hmm, not sure! Could you push the changes to this branch so that I can take a look?

Comment (pooja-muchandikar) [reply test/e2e/specs/editor/blocks/buttons.spec.js]: Hi,
I have pushed the code but the CI failed on same line.

Screenshot 2022-06-16 at 12 23 46 PM

--

Can we use this 👇  selector instead?

```
await expect(
      page.locator( 'a[href="https://example.com/"]' )
).toBeFocused();
```

Comment (kevin940726) [reply test/e2e/specs/editor/blocks/buttons.spec.js]: Oh, I think the URL input automatically trims the `https://` part off, so it's just `wordpress.org` now.

I think we can do this instead:

```js
await expect( page.locator( 'role=link[name=/^example\\.com/]' ) ).toBeFocused();
```

Comment (pooja-muchandikar) [reply test/e2e/specs/editor/blocks/buttons.spec.js]: Addressed the feedbacks.

File: test/e2e/specs/editor/blocks/buttons.spec.js [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+test.describe( 'Buttons', () => {
+	test.beforeEach( async ( { admin } ) => {
+		await admin.createNewPost();
+	} );
+
+	test( 'has focus on button content', async ( { editor, page } ) => {
+		await editor.insertBlock( { name: 'core/buttons' } );
+		await page.keyboard.type( 'Content' );
+
+		//Check the content
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: Let's fix the styling of the comments or just remove them if they are not helpful.

```suggestion
		// Check the content.
```

File: test/e2e/specs/editor/blocks/buttons.spec.js [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+	test( 'has focus on button content (slash inserter)', async ( {
+		editor,
+		page,
+	} ) => {
+		await page.click( 'role=button[name="Add default block"i]' );
+		await page.keyboard.type( '/buttons' );
+		await page.keyboard.press( 'Enter' );
+		await page.keyboard.type( 'Content' );
+
+		//Check the content
```
Comment (kevin940726) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: Ditto.

**Review by kevin940726 — APPROVED** [review]
Just some styling nitpicks left and then we can merge it! Thanks for the hard work!

File: test/e2e/specs/editor/blocks/buttons.spec.js (hunk 9 above)
Comment (pooja-muchandikar) [reply test/e2e/specs/editor/blocks/buttons.spec.js]: Changes Done :)

**Comment by kevin940726** [comment]
Not sure why the e2e tests are failing. Seems unrelated though. Could you help rebase it to the latest trunk and try again?

**Comment by pooja-muchandikar** [comment]
Hi @kevin940726,

I rebased the latest trunk and pushed the changes, but seems like the CI is failing. But not on the test case I worked on. 

image

---

**Comment by kevin940726** [comment]
Yeah, that's unrelated and is being tracked in https://github.com/WordPress/gutenberg/issues/37976 👍

**Comment by pooja-muchandikar** [comment]
Ok great!, any other changes required for this PR?

**Comment by kevin940726** [comment]
Nope, looks perfect to me! Tests are failing but look unrelated to this PR so I think we can merge this 👍

@talldan, could you help merge it?

**Comment by talldan** [comment]
I re-ran the tests, and the navigation block tests are still failing. They're unrelated to this PR so I'll merge it, and I'll also look at migrating the navigation block tests to playwright this week. Maybe that'll improve stability.

**Comment by pooja-muchandikar** [comment]
Thank you @kevin940726 @talldan 🙇🏻‍♀️

**Pull request merged**